addr_manager_ds.go

```diff
@@ -0,0 +1,267 @@
+		}
+		keys = append(keys, key)
+
+		if ttl <= 0 {
+			mgr.ds.Delete(key)
+			continue
+		}
+		if has, err := mgr.ds.Has(key); err != nil || !has {
+			mgr.subsManager.BroadcastAddr(p, addr)
+		}
+		if err := mgr.ds.Put(key, addr.Bytes()); err != nil {
+			log.Error(err)
+		}
+	}
+	mgr.ttlManager.setTTLs(keys, ttl)
+}
+
+func (mgr *DatastoreAddrManager) UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL time.Duration) {
+	prefix := ds.NewKey(p.Pretty())
+	mgr.ttlManager.updateTTLs(prefix, oldTTL, newTTL)
+}
+
+func (mgr *DatastoreAddrManager) Addrs(p peer.ID) []ma.Multiaddr {
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []ma.Multiaddr{}
+	}
+
+	var addrs []ma.Multiaddr
+	for result := range results.Next() {
+		addrbytes := result.Value.([]byte)
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
+			continue
+		}
+		addrs = append(addrs, addr)
+	}
+
+	return addrs
+}
+
+func (mgr *DatastoreAddrManager) Peers() []peer.ID {
+	q := query.Query{}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []peer.ID{}
+	}
+
+	idset := make(map[peer.ID]struct{})
+	for result := range results.Next() {
+		key := ds.RawKey(result.Key)
+		id, err := peerIDFromKey(key)
+		if err != nil {
+			continue
+		}
+		idset[id] = struct{}{}
+	}
+
+	ids := make([]peer.ID, 0, len(idset))
+	for id := range idset {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+func (mgr *DatastoreAddrManager) AddrStream(ctx context.Context, p peer.ID) <-chan ma.Multiaddr {
+	initial := mgr.Addrs(p)
+	return mgr.subsManager.AddrStream(ctx, p, initial)
+}
+
+func (mgr *DatastoreAddrManager) ClearAddrs(p peer.ID) {
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	for result := range results.Next() {
+		mgr.ds.Delete(ds.NewKey(result.Key))
+	}
+	mgr.ttlManager.clear(ds.NewKey(p.Pretty()))
+}
+
+// ttlmanager
+
+type ttlentry struct {
+	TTL       time.Duration
+	ExpiresAt time.Time
+}
+
+type ttlmanager struct {
+	sync.Mutex
+	entries map[ds.Key]*ttlentry
+	ctx     context.Context
+	cancel  context.CancelFunc
+	ticker  *time.Ticker
+	done    chan struct{}
+	ds      ds.Datastore
+}
+
+func newTTLManager(parent context.Context, d ds.Datastore, tick time.Duration) *ttlmanager {
+	ctx, cancel := context.WithCancel(parent)
+	mgr := &ttlmanager{
+		Mutex:   sync.Mutex{},
+		entries: make(map[ds.Key]*ttlentry),
+		ctx:     ctx,
+		cancel:  cancel,
+		ticker:  time.NewTicker(tick),
+		ds:      d,
+		done:    make(chan struct{}),
+	}
+
+	go func() {
+		for {
+			select {
+			case <-mgr.ctx.Done():
+				mgr.ticker.Stop()
+				mgr.done <- struct{}{}
+				return
+			case <-mgr.ticker.C:
+				mgr.tick()
+			}
+		}
+	}()
+
+	return mgr
+}
+
+func (mgr *ttlmanager) stop() {
+	mgr.cancel()
+	<-mgr.done
+}
+
+// For internal use only
```

Comment:
Internal to what the exactly? It is already internal to the package ;) I am guessing you mean for use by only the `DatastoreAddrManager`.